examples/blake2f.rs

```diff
@@ -0,0 +1,1377 @@
+            }
+
+            // check v_vec v_vec.next
+            for &v in v_vec.iter().take(12) {
+                ctx.transition(eq(v, v.next()));
+            }
+            ctx.transition(eq(
+                select(f, final_bits_sum_value, v_vec[14]),
+                v_vec[14].next(),
+            ));
+            ctx.transition(eq(v_vec[15], v_vec[15].next()));
+            // check h_vec h_vec.next
+            for &h in h_vec.iter() {
+                ctx.transition(eq(h, h.next()));
+            }
+            // check m_vec m_vec.next
+            for &m in m_vec.iter() {
+                ctx.transition(eq(m, m.next()));
+            }
+
+            ctx.constr(eq(round, 0));
+            ctx.transition(eq(round, round.next()));
+        });
+
+        ctx.wg(move |ctx, inputs: PreInput<F>| {
+            ctx.assign(round, inputs.round);
+            ctx.assign(t0, inputs.t0);
+            ctx.assign(t1, inputs.t1);
+            ctx.assign(f, inputs.f);
+            for (&q, &v) in wg_v_vec.iter().zip(inputs.v_vec.iter()) {
+                ctx.assign(q, v)
+            }
+            for (&q, &v) in wg_h_vec.iter().zip(inputs.h_vec.iter()) {
+                ctx.assign(q, v)
+            }
+            for (&q, &v) in wg_m_vec.iter().zip(inputs.m_vec.iter()) {
+                ctx.assign(q, v)
+            }
+            for (q_vec, v_vec) in wg_h_split_4bits_vec
+                .iter()
+                .zip(inputs.h_split_4bits_vec.iter())
+            {
+                for (&q, &v) in q_vec.iter().zip(v_vec.iter()) {
+                    ctx.assign(q, v)
+                }
+            }
+            for (q_vec, v_vec) in wg_m_split_4bits_vec
+                .iter()
+                .zip(inputs.m_split_4bits_vec.iter())
+            {
+                for (&q, &v) in q_vec.iter().zip(v_vec.iter()) {
+                    ctx.assign(q, v)
+                }
+            }
+            for (q_vec, v_vec) in wg_t_split_4bits_vec
+                .iter()
+                .zip(inputs.t_split_4bits_vec.iter())
+            {
+                for (&q, &v) in q_vec.iter().zip(v_vec.iter()) {
+                    ctx.assign(q, v)
+                }
+            }
+            for (q_vec, v_vec) in wg_iv_split_4bits_vec
+                .iter()
+                .zip(inputs.iv_split_4bits_vec.iter())
+            {
+                for (&q, &v) in q_vec.iter().zip(v_vec.iter()) {
+                    ctx.assign(q, v)
+                }
+            }
+            for (q_vec, v_vec) in wg_final_split_bits_vec
+                .iter()
+                .zip(inputs.final_split_bits_vec.iter())
+            {
+                for (&q, &v) in q_vec.iter().zip(v_vec.iter()) {
+                    ctx.assign(q, v)
+                }
+            }
+        })
+    });
+
+    let blake2f_g_step_vec: Vec<StepTypeWGHandler<F, _, _>> = (0..MIXING_ROUND)
+        .map(|r| {
+            ctx.step_type_def("blake2f_g_step", |ctx| {
```

Comment:
@10to4 perhaps use format! so the name of the step type contains `r` value?

Reply:
Whether we can't contain `r` value while using the function below?

```rust
impl From<&'static str> for StepTypeDefInput {
    fn from(s: &'static str) -> Self {
        StepTypeDefInput::String(s)
    }
}
```

It seems we need to add a new function `fn from(s: &str) -> Self ` or `fn from(s: String) -> Self` function for StepTypeDefInput.

Reply:
I don't understand exactly but you can add the `From for StepTypeDefInput` for the type that format! returns.